### ISSUE TITLE
CFP: Option to share load balancer IP by default

### ISSUE THREAD

**Issue by stevefan1999-personal** [cilium/cilium#30703]
## Cilium Feature Proposal

**Describe the feature you'd like**

Let us have an option to share the load balancer IP by default, rather than having to specify `lb-ipam-sharing-key` for every load balancer services.

**Is your proposed feature related to a problem?**

I'm running Talos Linux over multiple VPSes over the world. I'm using Cilium so that I can squeeze as much performance as possible. Rather than having a software defined network, most VPSes are just KVM and only have a single public IP hardcoded to the server. Examples: Amazon Lightsail, RackNerd, Hetzner, LayerStack (they claimed to be a cloud provider, but they don't even have software defined storage!)

In the past I have used [k3s](https://github.com/k3s-io/k3s) that included an interesting load balancer called Klipper, which uses a DaemonSet and a host port and then use iptables rule to SNAT the incoming connection into the Kubernetes world (yes, it is quite dangerous as it overwrites the host iptables), which means any servers not having those fancy Layer 3 BGP services or Layer 2 ARP network, can have an out of the box load balancer that is decently working to serve the internet traffic, unless you decided to screw up the iptables.

However, that load balancer doesn't seems to be public for other distribution to use, and due to a lot of restrictions that k3s imposes (such as the lack of ZFS snapshotter at that time), I have to give it up and turn around for something else. That's when I discovered Calico (and eventually Cilium) and after some time I decided to go for k0s.

Klipper isn't available for k0s, so I used OpenELB/Porter instead. Although OpenELB also suggest you use their fancy BGP/ARP features, they provided something similar in spirit of Klipper, albeit in alpha status. Their work in progress can be tracked [here](https://github.com/openelb/openelb/issues/214), although it has been staling for some time.

Sadly, after using Cilium, the OpenELB node proxy solution doesn't work anymore, as Cilium relies on eBPF and not iptables/IPVS anymore either, so their hack is rendered useless (in fact, that thing only works specifically under kube-proxy iptables mode). However, it has come to my mind that I can allocate a load balancer pool in Cilium, with the external IP of my VPSes, given my two VPSes have public IPs of "172.18.0.1" and "193.162.1.1" for example:
```
apiVersion: "cilium.io/v2alpha1"
kind: CiliumLoadBalancerIPPool
metadata:
  name: "default"
spec:
  blocks:
  - cidr: "172.18.0.1/32"
  - cidr: "193.162.1.1/32"
```
This worked out of the box with no surprises. Then I create a load balancer for a service:
```
apiVersion: v1
kind: Service
metadata:
  name: nginx-deployment
  namespace: default
  labels:
    app: nginx
spec:
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 80
  selector:
    app: nginx
  type: LoadBalancer
```
And given the luck I can actually access my nginx service on either `172.18.0.1:8080` or `193.162.1.1:8080`! However, as the load balancer IP is allocated exclusively, we have to add a little more:
```
apiVersion: v1
kind: Service
metadata:
  name: nginx-deployment1
  namespace: default
  labels:
    app: nginx
  annotations:
    io.cilium/lb-ipam-sharing-key: default
spec:
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 80
  selector:
    app: nginx
  type: LoadBalancer
```
This will allow the load balancer IP to be shared (however, that feature is semi-broken due to #30700), but we have to add the `io.cilium/lb-ipam-sharing-key: default` to all services. And it is not very possible sometimes, for example, I don't know if I can add annotations to the autogenerated LoadBalancer service for [Agones](https://agones.dev/site/) game server allocation, and some Helm chart author may also have not been aware of letting people add annotations to the services.

So maybe we should have a switch that let this sharing behavior to be the default, unless fields like `io.cilium/lb-ipam-sharing-key` are manually specified. 

It is possible to do this via a third-party operator and a admission controller (more specifically a mutation webhook on the Service kind), which I'm writing a prototype for this kind of use, that snoops the annotation fields and check the presence for the fields like `io.cilium/lb-ipam-sharing-key` and such, and if it is not, we add in preset values to the annotation map and thus getting a similar result. 

The operator will also collect the external IPs of the nodes in the cluster, and autogenerate a CiliumLoadBalancerIPPool based on those IPs alone, using the `/32` trick. This however is tedious and can be bug prone, so I want to see if Cilium can implement this onto itself.

**Comment by borkmann** [comment]
Cc @dylandreimerink

**Comment by dylandreimerink** [comment]
Recently [this PR](https://github.com/cilium/cilium/pull/30038) added something called nodeipam which I believe does what you are describing. That feature does require you to set a specific LB class, so that might not fix your Agones situation, but I would argue that that is up to them.

Additionally, all of this sounds to me like a use case that a nodeport service can also fix (given you config the control plane to allow your port ranges).

If the above options don't satisfy your request then I would like to hear why not and some feedback from the community on expected adoption. I am not yet convinced that the benefits of adding this feature outweigh the added complexity and maintenance burden.

**Comment by stevefan1999-personal** [comment]
Closing as @dylandreimerink has suggested. However, IIRC the feature is not released in 1.15.0 yet

**Comment by dylandreimerink** [comment]
Yea, nodeipam was merged on Jan 11th, after the branch off for v1.15 so it will be in v1.16 and its release candidates. I believe `io.cilium/node` is also used in other places than the lb class which is why a grep may find something.
